compiler/include/PassManager.h

```diff
@@ -0,0 +1,138 @@
+// A PassT is a Pass that runs on things of type T and can return results of ResultType
+//
+// The default run behavior provides a common pattern that many passes
+// exhibit, along with providing a place for an idempotence check to
+// be used in the future. A PassT processes a single item of type T so
+// that the caller can drive the execution.
+//
+// There might be a useful baseclass for this that has more general `run` method
+// for instance, but for now having it templated over its input type is too useful.
+//
+// Run could equally be spelled operator() but I find the name more clear.
+// I expect most of our uses today to be de-virtualized (thinking of perf).
+//
+// Some passes accumulate some result after the whole pass is done,
+// which can be returned with getResult. This might be storing some
+// info in a map for instance. Future work could extend a pass to be
+// combinable (operator+ maybe) so that parallel execution could
+// divide, conquer, and combine the results.
+//
+//
+template <typename T, typename ResultType = void> class PassT {
+ public:
+  virtual ~PassT() = default;
+
+  virtual bool shouldProcess(T x) {
+    (void)x;
+    return true;
+  }
+
+  virtual void process(T x) {
+    // do nothing
+  }
+
+  // main driver method, provides
+  virtual void run(T x) {
+    if (shouldProcess(x)) {
+      process(x);
+    }
+  };
+
+  virtual ResultType getResult() {
+    // return nothing
+  }
+};
+
+template <typename T> using PassTList = std::vector<std::unique_ptr<PassT<T>>>;
+
+//
+// PassManager runs passes over some collection of objects
+//
+// Today, we only have "immediate mode" execution of a pass or group
+// of passes that operate on the same type. (as opposed to building up
+// something like an execution graph that is later executed) because
+// it still gives us the benefit of having the passmanager drive
+// execution (so that sequential passes that operate over the same
+// type can be chained (where possible)), but without the
+// complexity. Part of the implementation difficulty with that
+// approach today is that we have passes that operate on a load of
+// different types and building up a typesafe graph of those is more
+// trouble than worth at this point. One can imagine this would change
+// when we have fewer pass types, like say Global, Module, Type, and
+// Function. Its not clear to me at this point we can flip the switch
+// today so that eg. a pass over CallExpr is actually a Function pass
+// with an adapter to visit every CallExpr without a performance hit.
+//
+class PassManager {
+ private:
+  // Run pass over many and return it's results (if any)
+  template <typename T, typename R, typename Container>
+  R runPass(PassT<T, R>& pass, const Container& xs) {
+    for (auto& x : xs) {
+      pass.run(x);
+    }
+    return pass.getResult();
+  }
+
+ public:
+  // We take pass by && in these for the common case where the caller
+  // sinks their argument; like runPass(myPassName(), ...);
```

Comment:
I left it that way since the ones over PassTList are move as well. I can add an overload in the future. I'm not still certain on what the lifetimes of Passes should be and how PassManager will/won't own them so this could easily change in the future.
https://isocpp.github.io/CppCoreGuidelines/CppCoreGuidelines#f18-for-will-move-from-parameters-pass-by-x-and-stdmove-the-parameter